### ISSUE TITLE
Lint expects a camel-case identifier for something that is not even an identifier (i.e. in patterns)

### ISSUE THREAD

**Issue by abelbraaksma** [fsprojects/FSharpLint#188]
### Description

In left-hand declaration patterns, Lint will complain if you use a discriminated union and it'll expect a camel-case identifier, which is virtually impossible because DU's must start with a capital letter.
### Repro steps

The following code shows this behavior:

```
type WithCamel = YesCamel

module Camels =
    let SomeCamel WithCamel.YesCamel = 12    
```
### Expected behavior

No warning whatsoever (though my own, and F#'s style guide's, preference is to have camel-case for all let-bindings, public and otherwise, the default of FSharpLint seems to be to expect a capital).
### Actual behavior

A warning is raised:

> Lint: Expected a camel case identifier, found 'YesCamel'.

Or with F# Power Tools, as in the screenshot:

![Camel Case Warning](http://i.imgur.com/Enw2W1k.png)
### Known workarounds

None (or remove the warning for the whole file or block, but that removes the ability for lint-testing the rest).
### Related information

VS2015, Update 3, with Visual F# Power Tools 2.5.2.

**Comment by duckmatt** [comment]
Thanks for the report, I'll take a look into this

> (though my own, and F#'s style guide's, preference is to have camel-case for all let-bindings, public and otherwise, the default of FSharpLint seems to be to expect a capital).

For which rule is this the case? public let bindings can be either PascalCase or camelCase according to the style guide, often being PascalCase when consumed by other .NET languages

**Comment by abelbraaksma** [comment]
> For which rule is this the case? public let bindings can be either PascalCase or camelCase according to the style guide, often being PascalCase when consumed by other .NET languages

@duckmatt, it is in the [Component Design](http://fsharp.org/specs/component-design-guidelines/#suggest-using-camelcase-for-class-bound-expression-bound-and-pattern-bound-values-and-functions).

It also says:

> In large assemblies and implementation files, PascalCase is **sometimes** used to indicate major internal routines.

The emphasis on _sometimes_ is mine. To the best of my knowledge, all public F# functions in the F# Core are camel-case (`Seq.ofArray`, `Option.map` etc) and sometimes lowercase (`failwith`, `failwithf`) and where there are wrappers on LINQ, the names are camel-cased (`headOrDefault`). Most F# documentation found online shows let-bindings (public and private) as camel-case. The MSDN documentation often writes about such members "if you use this method outside F#, use MemberName, i.e. the PascalCase version", which, if needed, I achieve by using `CompiledNameAttribute`.

Bottom line, I think, is that let- and val-bindings are always camelCase, static class members are camelCase, instance members are camelCase for private, and PascalCase for public. I am not sure this is spelled out anywhere so explicitly and I may be wrong, but this is what I've deducted over the years and what is commonly used in F# projects.

**Comment by duckmatt** [comment]
I agree with what you're saying, personally I don't tend to use PascalCase for any let bindings. Using PascalCase to indicate major internal routines also seems a bit odd to me; where the default setting has come from is in section `3.1 Naming Conventions`, for the construct: `let values (external)`, it suggests `camelCase or PascalCase`

**Comment by duckmatt** [comment]
Added a [regression test](https://github.com/fsprojects/FSharpLint/blob/1d6f87e9feef42ee0e8fac7d343087547240939a/tests/FSharpLint.Core.Tests/Rules/TestNameConventionRules.fs#L1373) for this case, thanks for reporting
